docs/textsplitter.md

```diff
@@ -0,0 +1,284 @@
+# RAG Chat: Text splitting algorithm overview
+
+This document explains the chunking logic implemented in the [data ingestion pipeline](./data_ingestion.md). The [splitter module](../app/backend/prepdocslib/textsplitter.py) contains both a `SimpleTextSplitter` (used only for JSON files) and a `SentenceTextSplitter` (used for all other formats). This document focuses on the `SentenceTextSplitter` since its approach is far more complicated, and it can be difficult to follow the code.
+
+* [High-level overview](#high-level-overview)
+* [Splitting algorithm](#splitting-algorithm)
+* [Recursive handling of oversized spans](#recursive-handling-of-oversized-spans)
+* [Cross-page boundary repair](#cross-page-boundary-repair)
+* [Chunk normalization](#chunk-normalization)
+* [Semantic overlap](#semantic-overlap)
+* [Examples](#examples)
+
+## High-level overview
+
+The `SentenceTextSplitter` is designed to:
+
+1. Produce semantically coherent chunks that align with sentence boundaries.
+2. Respect a maximum token count per chunk (hard limit of 500 tokens) plus a soft character length guideline (default 1,000 characters with a 20% overflow tolerance for merges / normalization). Size limit does not apply to figure blocks (chunks containing a `<figure>` may exceed the token limit; figures are never split).
+3. Keep structural figure placeholders (`<figure>...</figure>`) atomic: never split internally and always attach them to preceding accumulated text if any exists.
+4. Repair mid‑sentence page breaks when safe via merge or fragment shift heuristics while enforcing token + soft character budgets.
+5. Avoid empty outputs or unclosed figure tags.
+6. Perform a light normalization pass (trim only minimal leading/trailing whitespace that would cause small overflows; do not modify figure chunks).
+
+The splitter includes these components:
+
+* Segmentation based on sentence-ending punctuation ( `. ! ?` plus CJK equivalents ).
+* An internal accumulator that appends sentence‑like spans until the next addition would breach character or token limits, then flushes a chunk. Default hard token cap is 500 per chunk.
+* Recursive subdivision of oversized individual spans using a boundary preference order:
+    1. Sentence-ending punctuation near the midpoint (scan within the central third of the span).
+    2. If no sentence boundary is found, a word break (space / punctuation from a configured list) near the midpoint to avoid mid‑word cuts.
+    3. If neither boundary type is found, a symmetric 10% overlap midpoint split (duplicated region appears at the end of the first part and the start of the second) preserves continuity.
```

Comment:
symmetric 10% overlap midpoint split - perhaps we just say "default to a simpler midpoint split"?